### PR TITLE
Handle file URLs

### PR DESCRIPTION
**Pull request by alan-knight**

### CODE REVIEW

File: _test/lib/library.dart [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// An library that we can import.
```
Comment (natebosch) [inline _test/lib/library.dart]: [nit] `An` -> `A`

Also, should we add a library directive so this doesn't appear to be a doc comment on a variable?

File: dwds/_testPackage/pubspec.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+name: _testPackage
+version: 1.0.0
+description: >-
+  A fake package used for testing imports. Imports _test.
+publish_to: none
+
+environment:
+  sdk: ">=2.3.2-dev.0.1 <3.0.0"
```
Comment (natebosch) [inline dwds/_testPackage/pubspec.yaml]: [nit] Why so specific on the lower bound?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 3]
```diff
@@ -14,6 +15,43 @@ class DartUri {
   /// 'packages/path/src/utils.dart'.
   final String serverPath;
 
+  /// The directory in which we're running.
+  ///
+  /// We store this here because for tests we may want to act as if we're
```
Comment (natebosch) [inline dwds/lib/src/utilities/dart_uri.dart]: Would it be sensible to change the working directory entirely during tests?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 4]
```diff
@@ -14,6 +15,43 @@ class DartUri {
...
+  /// running in the directory of a target package, even if the current
+  /// directory of the tests is actually the main dwds directory.
+  static String currentDirectory = p.current;
+
+  /// Load the .packages file associated with the running application so we can
+  /// resolve file URLs into package: URLs appropriately.
+  static Future<void> loadPackageConfig() async {
```
Comment (natebosch) [inline dwds/lib/src/utilities/dart_uri.dart]: Should this be private?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 5]
```diff
@@ -14,6 +15,43 @@ class DartUri {
...
+  }
+
+  /// The way we resolve file: URLs into package: URLs
+  static SyncPackageResolver _packageResolver;
+
+  /// All of the known libraries, indexed by their absolute file URL.
+  static final Map<String, String> _libraryNamesByPath = {};
+
+  /// Record all of the libraries, indexed by their absolute file: URL.
+  static Future<Null> noteLibraries(Iterable<String> libraryUris) async {
```
Comment (natebosch) [inline dwds/lib/src/utilities/dart_uri.dart]: `Future<void>`

File: dwds/test/dart_uri_file_uri_test.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:io';
```
Comment (natebosch) [inline dwds/test/dart_uri_file_uri_test.dart]: [nit] Can we put a blank line below the copyright?

File: dwds/test/dart_uri_file_uri_test.dart [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import 'package:path/path.dart' as p;
+
+import 'package:dwds/src/utilities/dart_uri.dart';
+import 'package:test/test.dart';
```
Comment (natebosch) [inline dwds/test/dart_uri_file_uri_test.dart]: [nit] import sections aren't grouped properly. The `package:` imports` should all be in the same block.

File: dwds/lib/src/utilities/dart_uri.dart [hunk 8]
```diff
@@ -14,6 +15,43 @@ class DartUri {
...
+        var uriPath = Uri.parse(uri).path;
+        // The Uri's path will be absolute, remove the leading slash.
+        var libraryPath = p.join(currentDirectory, uriPath.substring(1));
+        _libraryNamesByPath[Uri.file(libraryPath).toString()] = uri;
+      }
+      if (uri.startsWith('package:')) {
+        var libraryPath = _packageResolver.resolveUri(uri);
+        _libraryNamesByPath['$libraryPath'] = uri;
+      }
+    }
```
Comment (jakemac53) [inline dwds/lib/src/utilities/dart_uri.dart]: nit: throw if we don't hit the other cases

File: dwds/_testPackage/pubspec.yaml (hunk 2 above)
Comment (alan-knight) [reply dwds/_testPackage/pubspec.yaml]: Actually that comment is outdated now. Updated everything to be 2.5.0 so we consistently get org-dartlang-app: URLs.

File: dwds/lib/src/utilities/dart_uri.dart [hunk 9]
```diff
@@ -35,20 +35,31 @@ class DartUri {
...
+  }
+
+  /// Record the library represented by package: or org-dartlang-app: uris
+  /// indexed by absolute file: URI.
+  static void recordAbsoluteUri(String libraryUri) {
+    var uri = Uri.parse(libraryUri);
+    if (uri.scheme == 'dart' ||
+        (uri.scheme == '' && !uri.path.endsWith('.dart'))) {
+      /// We ignore dart: libraries, and non-Dart libraries referenced by path.
```
Comment (jakemac53) [inline dwds/lib/src/utilities/dart_uri.dart]: What are these?

File: dwds/test/dart_uri_file_uri_test.dart [hunk 10]
```diff
@@ -50,4 +54,11 @@ void main() {
     var dartUri = DartUri('$testLib');
     expect(dartUri.serverPath, 'packages/_test/library.dart');
   });
+
+  test('Combining Windows current directory with URI path', () {
```
Comment (jakemac53) [inline dwds/test/dart_uri_file_uri_test.dart]: This is a weird test to live here? It is only testing package:path as far as I can tell?

File: fixtures/_test/.vscode/launch.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (jakemac53) [inline fixtures/_test/.vscode/launch.json]: This should probably be added to your global git ignores - I assume we don't intend to submit it here?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 12]
```diff
@@ -14,6 +15,54 @@ class DartUri {
...
+  static Future<void> _loadPackageConfig() async {
+    _packageResolver ??= await SyncPackageResolver.loadConfig(
+        p.toUri(p.join(currentDirectory, '.packages')));
+  }
+
+  /// The way we resolve file: URLs into package: URLs
+  static SyncPackageResolver _packageResolver;
+
+  /// All of the known libraries, indexed by their absolute file URL.
+  static final Map<String, String> _libraryNamesByPath = {};
```
Comment (grouma) [inline dwds/lib/src/utilities/dart_uri.dart]: I don't see this getting cleared out. Shouldn't it though?

File: dwds/test/dart_uri_file_uri_test.dart [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+String get dwdsDir => Directory.current.absolute.path;
+
+/// The directory for the general _test package.
+String get testDir => p.join(p.dirname(dwdsDir), 'fixtures', '_test');
+
+/// The directory for the _testPackage packagwe (contained within dwds), which imports _test.
+String get testPackageDir => context.workingDirectory;
+
+/// This tests converting file Uris into our internal paths.
```
Comment (grouma) [inline dwds/test/dart_uri_file_uri_test.dart]: This probably should be a regular comment?

File: dwds/lib/src/utilities/dart_uri.dart [hunk 14]
```diff
@@ -14,6 +15,54 @@ class DartUri {
...
+    }
+  }
+
+  /// Record the library represented by package: or org-dartlang-app: uris
+  /// indexed by absolute file: URI.
+  static void recordAbsoluteUri(String libraryUri) {
+    var uri = Uri.parse(libraryUri);
+    if (uri.scheme == 'dart' ||
+        (uri.scheme == '' && !uri.path.endsWith('.dart'))) {
+      /// We ignore dart: libraries, and non-Dart libraries referenced by path.
```
Comment (natebosch) [inline dwds/lib/src/utilities/dart_uri.dart]: 2 slashes

**Pull request merged**